File: nodejs/src/executors/crawlExecutor.js
const util = require('util');
const http = require('http');
const https = require('https');
const urlLib = require('url');
const EventEmitter = require('events');
const executorUtil = require('./executorUtil');

function CrawlExecutor() {
    EventEmitter.call(this);

    function onRegistered(driver, executorInfo, frameworkInfo, slaveInfo) {
        console.log("Registered executor on host " + slaveInfo.hostname);
    }

    function onLaunchTask(driver, taskInfo) {
        console.log("Launching crawl task '" + taskInfo.task_id.value + "'...");
        executorUtil.sendTaskRunningStatus(driver, taskInfo.task_id);

        var url = taskInfo.data.toUTF8();

        var httpx = getHttpObject(url);
        if (!httpx) {
            console.log("Unrecognized url protocol: " + url);
            executorUtil.sendTaskFinishedStatus(driver, taskInfo.task_id);
            return;
        }

        httpx.get(url, function (response) {
            var links = [];
            response.setEncoding('utf8');

            response.on('data', function (chunk) {
                links = links.concat(parseLinks(chunk));
            });
            response.on('end', function () {
                sendCrawlResultMessage(driver, url, links);
                executorUtil.sendTaskFinishedStatus(driver, taskInfo.task_id);
            });
        }).on('error', function (error) {
            console.log("Error during crawl operation: " + error);
            executorUtil.sendTaskErrorStatus(driver, taskInfo.task_id);
        });
    }

    function onError(driver, message) {
        console.log("Error: " + message);
    }

    function sendCrawlResultMessage(driver, url, links) {
        var message = new {
            type: "CrawlResult",
            body: {
                url: url,
                links: links
            }
        };

        driver.sendFrameworkMessage(JSON.stringify(message));
    }

    function getHttpObject(url) {
        var parsed = urlLib.parse(url);
        switch (parsed.protocol) {
            case "https:":
                return https;
            case "http:":
                return http;
            default :
                return null;
        }
    }

    function parseLinks(content) {
        var regex = /<a[^>]+href=[\"']?([^\"\'>]+)[\"\']?[^>]*>.+?<\/a>/gi;
        var links = [];
        var array;
        while ((array = regex.exec(content)) !== null) {
            var link = array[1];
            links.push(link);
        }

        return links;
    }

    this.on("registered", onRegistered);
    this.on("launchTask", onLaunchTask);
    this.on("error", onError);
}

util.inherits(CrawlExecutor, EventEmitter);

module.exports = CrawlExecutor;

File: csharp/src/Rendler/Executors/CrawlExecutor.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using mesos;
using mesosclr;
using Rendler.Executors.Messages;
using System.Net.Http;

namespace Rendler.Executors
{
    internal class CrawlExecutor : ExecutorBase
    {
        private static readonly Regex ExtractLinksRegex = new Regex("<a[^>]+href=[\"']?(?<link>[^\"'>]+)[\"']?[^>]*>(.+?)</a>", RegexOptions.Compiled | RegexOptions.IgnoreCase);

        public override void Registered(IExecutorDriver driver, ExecutorInfo executorInfo, FrameworkInfo frameworkInfo, SlaveInfo slaveInfo)
        {
            Console.WriteLine($"Registered executor on '{slaveInfo.hostname}'.");
        }

        public override void LaunchTask(IExecutorDriver driver, TaskInfo taskInfo)
        {
            Console.WriteLine($"Launching crawl task '{taskInfo.task_id.value}'...");

            Task.Factory.StartNew(async () =>
            {
                try
                {
                    await RunTask(driver, taskInfo);
                }
                catch (Exception e)
                {
                    Console.WriteLine($"Exception during crawl operation: {e}");
                    driver.SendTaskErrorStatus(taskInfo.task_id);
                }
            });
        }

        private static async Task RunTask(IExecutorDriver driver, TaskInfo taskInfo)
        {
            driver.SendTaskRunningStatus(taskInfo.task_id);

            var url = Encoding.UTF8.GetString(taskInfo.data);

            var htmlContent = await GetUrlContent(url);
            if (htmlContent != null)
            {
                var links = ExtractLinks(htmlContent);
                links = links
                    .Select(x => x.ToLower())
                    .Distinct(StringComparer.CurrentCultureIgnoreCase);

                if (links.Any())
                    SendCrawlResultMessage(driver, url, links.ToArray());
            }

            driver.SendTaskFinishedStatus(taskInfo.task_id);
        }

        private static IEnumerable<string> ExtractLinks(string htmlContent)
        {
            var match = ExtractLinksRegex.Match(htmlContent);
            while (match.Success)
            {
                yield return match.Groups["link"].Value.Trim();
                match = match.NextMatch();
            }
        }

        private static async Task<string> GetUrlContent(string url)
        {
            using (var client = new HttpClient())
            {
                client.DefaultRequestHeaders.Add("X-PoweredBy", "minions");

                try
                {
                    return await client.GetStringAsync(url);
                }
                catch (WebException e)
                {
                    Console.WriteLine($"Error fetching url '{url}'; Error: {e}");
                    return null;
                }
            }
        }

        private static void SendCrawlResultMessage(IExecutorDriver driver, string url, string[] links)
        {
            var message = new Message
            {
                Type = "CrawlResult",
                Body = JsonHelper.Serialize(new CrawlResultMessage
                {
                    Url = url,
                    Links = links
                })
            };

            driver.SendFrameworkMessage(JsonHelper.Serialize(message));
        }
    }
}
